File: Buddy/buddy-ios/Sources/buddy/Core/Services/RandomEventService.swift
import Foundation

enum RandomEvent: String, CaseIterable {
    case giftBox            // bonus coins
    case foundFood          // hunger boost free
    case visitor            // happiness boost
    case stormyMood         // happiness drop (drama)
    case gustOfWind         // hygiene drop
    case lazyMode           // energy drop

    var emoji: String {
        switch self {
        case .giftBox:    "🎁"
        case .foundFood:  "🍱"
        case .visitor:    "👋"
        case .stormyMood: "⛈"
        case .gustOfWind: "🌬"
        case .lazyMode:   "😴"
        }
    }

    var title: String {
        switch self {
        case .giftBox:    "Caja sorpresa"
        case .foundFood:  "Comida encontrada"
        case .visitor:    "Tu mascota recibió una visita"
        case .stormyMood: "Cambio de humor"
        case .gustOfWind: "Polvo en el ambiente"
        case .lazyMode:   "Día flojo"
        }
    }

    var detail: String {
        switch self {
        case .giftBox:    "+15 monedas"
        case .foundFood:  "+25 hambre"
        case .visitor:    "+20 felicidad"
        case .stormyMood: "-15 felicidad"
        case .gustOfWind: "-15 higiene"
        case .lazyMode:   "-20 energía"
        }
    }

    var isPositive: Bool {
        switch self {
        case .giftBox, .foundFood, .visitor: true
        default: false
        }
    }
}

struct RandomEventService {
    static func maybeTrigger(pet: Pet) -> RandomEvent? {
        // 8% chance per call
        guard Int.random(in: 0..<100) < 8 else { return nil }
        let event = RandomEvent.allCases.randomElement()!
        apply(event, to: pet)
        return event
    }

    private static func apply(_ event: RandomEvent, to pet: Pet) {
        switch event {
        case .giftBox:    CoinWallet.add(15)
        case .foundFood:  pet.stats.hunger = min(100, pet.stats.hunger + 25)
        case .visitor:    pet.stats.happiness = min(100, pet.stats.happiness + 20)
        case .stormyMood: pet.stats.happiness = max(0, pet.stats.happiness - 15)
        case .gustOfWind: pet.stats.hygiene = max(0, pet.stats.hygiene - 15)
        case .lazyMode:   pet.stats.energy = max(0, pet.stats.energy - 20)
        }
        pet.stats.clamp()
    }
}

File: Buddy/buddy-flutter/lib/core/services/random_event_service.dart
import 'dart:math';

import '../../models/pet.dart';
import '../persistence/coin_wallet.dart';

enum RandomEvent {
  giftBox,
  foundFood,
  visitor,
  stormyMood,
  gustOfWind,
  lazyMode;

  String get emoji => switch (this) {
        RandomEvent.giftBox    => '🎁',
        RandomEvent.foundFood  => '🍱',
        RandomEvent.visitor    => '👋',
        RandomEvent.stormyMood => '⛈',
        RandomEvent.gustOfWind => '🌬',
        RandomEvent.lazyMode   => '😴',
      };

  String get title => switch (this) {
        RandomEvent.giftBox    => 'Caja sorpresa',
        RandomEvent.foundFood  => 'Comida encontrada',
        RandomEvent.visitor    => 'Tu mascota recibió una visita',
        RandomEvent.stormyMood => 'Cambio de humor',
        RandomEvent.gustOfWind => 'Polvo en el ambiente',
        RandomEvent.lazyMode   => 'Día flojo',
      };

  String get detail => switch (this) {
        RandomEvent.giftBox    => '+15 monedas',
        RandomEvent.foundFood  => '+25 hambre',
        RandomEvent.visitor    => '+20 felicidad',
        RandomEvent.stormyMood => '-15 felicidad',
        RandomEvent.gustOfWind => '-15 higiene',
        RandomEvent.lazyMode   => '-20 energía',
      };

  bool get isPositive => switch (this) {
        RandomEvent.giftBox || RandomEvent.foundFood || RandomEvent.visitor => true,
        _ => false,
      };
}

class RandomEventService {
  static final _rng = Random();

  /// 8% de probabilidad por llamada. Si dispara, aplica el efecto y devuelve el evento.
  static Future<RandomEvent?> maybeTrigger(Pet pet) async {
    if (_rng.nextInt(100) >= 8) return null;
    final event = RandomEvent.values[_rng.nextInt(RandomEvent.values.length)];
    await _apply(event, pet);
    return event;
  }

  static Future<void> _apply(RandomEvent event, Pet pet) async {
    final s = pet.stats;
    switch (event) {
      case RandomEvent.giftBox:
        await CoinWallet.add(15);
        break;
      case RandomEvent.foundFood:
        s.hunger = (s.hunger + 25).clamp(0, 100);
        break;
      case RandomEvent.visitor:
        s.happiness = (s.happiness + 20).clamp(0, 100);
        break;
      case RandomEvent.stormyMood:
        s.happiness = (s.happiness - 15).clamp(0, 100);
        break;
      case RandomEvent.gustOfWind:
        s.hygiene = (s.hygiene - 15).clamp(0, 100);
        break;
      case RandomEvent.lazyMode:
        s.energy = (s.energy - 20).clamp(0, 100);
        break;
    }
    s.clamp();
    pet.stats = s;
  }
}
